Auto-detect C++ features and output .cpp by default

## Summary

When cnext parses headers that contain C++ features, it should automatically output `.cpp` instead of `.c` to prevent build errors.

## Problem

Currently, if you include a header with C++ features (like `AppConfig.h` with typed enums), cnext outputs a `.c` file. When compiled as C, this fails:

```
include/AppConfig.h:17:23: error: expected identifier or '(' before ':' token
```

The user must manually rename to `.cpp` or remember to use `--cpp`.

## Proposed Solution

Auto-detect C++ features during header parsing and switch output extension accordingly.

### Detection Triggers

- Typed enums: `enum Foo : uint8_t`
- Native `bool` without `#include <stdbool.h>`
- Classes or structs with methods
- `namespace` blocks
- Templates (`template<>`)
- C++ standard headers (`<iostream>`, `<vector>`, `<string>`, etc.)
- `extern "C"` blocks (indicates C++ context)

### Proposed CLI Behavior

```bash
# Auto-detect (default) - smart extension choice
cnext src/foo.cnx  → foo.cpp (if C++ detected) or foo.c (pure C)

# Explicit overrides for edge cases
cnext --c src/foo.cnx    → always output foo.c
cnext --cpp src/foo.cnx  → always output foo.cpp (existing flag)
```

### Benefits

1. **Better UX** - "just works" without remembering flags
2. **Prevents build errors** - the exact issue we hit with OSSM project
3. **Leverages existing parsing** - cnext already parses headers for type info, so it's already seeing these features
4. **Backwards compatible** - `--cpp` flag still works, and `--c` provides explicit control

## Context

This came up when testing cnext v0.1.7 with the OSSM embedded project. The generated `crc32.c` failed to compile because `AppConfig.h` uses C++11 typed enums. We had to manually rename to `.cpp` before discovering the `--cpp` flag.

--- a/src/index.ts
+++ b/src/index.ts
@@ -25,7 +25,8 @@ const packageJson = require("../package.json");
  * C-Next configuration file options
  */
 interface ICNextConfig {
-  outputExtension?: ".c" | ".cpp";
+  /** Issue #211: Force C++ output. Auto-detection may also enable this. */
+  cppRequired?: boolean;
   generateHeaders?: boolean;
   debugMode?: boolean;
   target?: string; // ADR-049: Target platform (e.g., "teensy41", "cortex-m0")
@@ -130,7 +131,7 @@ function showHelp(): void {
   console.log("  cnext.config.json, .cnext.json, .cnextrc");
   console.log("");
   console.log("Config example:");
-  console.log('  { "outputExtension": ".cpp" }');
+  console.log('  { "cppRequired": true }');
   console.log("");
   console.log("A safer C for embedded systems development.");
 }
@@ -188,7 +189,7 @@ async function runUnifiedMode(
   generateHeaders: boolean,
   preprocess: boolean,
   verbose: boolean,
-  outputExtension: ".c" | ".cpp",
+  cppRequired: boolean,
   noCache: boolean,
 ): Promise<void> {
   // Step 1: Expand directories to .cnx files
@@ -257,7 +258,7 @@ async function runUnifiedMode(
     generateHeaders,
     preprocess,
     defines,
-    outputExtension,
+    cppRequired,
     noCache,
   });
 
@@ -466,7 +467,7 @@ async function main(): Promise<void> {
   const includeDirs: string[] = [];
   const defines: Record<string, string | boolean> = {};
   let cliGenerateHeaders: boolean | undefined;
-  let cliOutputExtension: ".c" | ".cpp" | undefined;
+  let cliCppRequired: boolean | undefined;
   let preprocess = true;
   let verbose = false;
   let noCache = false;
@@ -483,7 +484,7 @@ async function main(): Promise<void> {
     } else if (arg === "--exclude-headers") {
       cliGenerateHeaders = false;
     } else if (arg === "--cpp") {
-      cliOutputExtension = ".cpp";
+      cliCppRequired = true;
     } else if (arg === "--no-preprocess") {
       preprocess = false;
     } else if (arg === "--no-cache") {
@@ -508,7 +509,7 @@ async function main(): Promise<void> {
 
   // Apply config defaults, CLI flags take precedence
   const generateHeaders = cliGenerateHeaders ?? config.generateHeaders ?? true;
-  const outputExtension = cliOutputExtension ?? config.outputExtension ?? ".c";
+  const cppRequired = cliCppRequired ?? config.cppRequired ?? false;
 
   // Unified mode - always use Project class with header discovery
   if (inputFiles.length === 0) {
@@ -525,7 +526,7 @@ async function main(): Promise<void> {
     generateHeaders,
     preprocess,
     verbose,
-    outputExtension,
+    cppRequired,
     noCache,
   );
 }

--- a/src/pipeline/Pipeline.ts
+++ b/src/pipeline/Pipeline.ts
@@ -56,6 +56,8 @@ class Pipeline {
   private headerGenerator: HeaderGenerator;
   private warnings: string[];
   private cacheManager: CacheManager | null;
+  /** Issue #211: Tracks if C++ output is needed (one-way flag, false → true only) */
+  private cppDetected: boolean;
 
   constructor(config: IPipelineConfig) {
     // Apply defaults
@@ -66,13 +68,16 @@ class Pipeline {
       defines: config.defines ?? {},
       preprocess: config.preprocess ?? true,
       generateHeaders: config.generateHeaders ?? true,
-      outputExtension: config.outputExtension ?? ".c",
+      cppRequired: config.cppRequired ?? false,
       parseOnly: config.parseOnly ?? false,
       debugMode: config.debugMode ?? false,
       target: config.target ?? "",
       collectGrammarCoverage: config.collectGrammarCoverage ?? false,
       noCache: config.noCache ?? false,
     };
+
+    // Issue #211: Initialize cppDetected from config (--cpp flag sets this)
+    this.cppDetected = this.config.cppRequired;
 
     this.symbolTable = new SymbolTable();
     this.preprocessor = new Preprocessor();
@@ -317,7 +322,20 @@ class Pipeline {
         this.symbolTable.restoreStructFields(cached.structFields);
         this.symbolTable.restoreNeedsStructKeyword(cached.needsStructKeyword);
         this.symbolTable.restoreEnumBitWidths(cached.enumBitWidth);
-        return; // Cache hit - skip parsing
+
+        // Issue #211: Still check for C++ syntax even on cache hit
+        // The detection is cheap (regex only) and ensures cppDetected is set correctly
+        if (file.type === EFileType.CHeader) {
+          const content = readFileSync(file.path, "utf-8");
+          if (detectCppSyntax(content)) {
+            this.cppDetected = true;
+          }
+        } else if (file.type === EFileType.CppHeader) {
+          // .hpp files are always C++
+          this.cppDetected = true;
+        }
+
+        return; // Cache hit - skip full parsing
       }
     }
 
@@ -347,6 +365,8 @@ class Pipeline {
     if (file.type === EFileType.CHeader) {
       this.parseCHeader(content, file.path);
     } else if (file.type === EFileType.CppHeader) {
+      // Issue #211: .hpp files are always C++
+      this.cppDetected = true;
       this.parseCppHeader(content, file.path);
     }
 
@@ -374,6 +394,8 @@ class Pipeline {
    */
   private parseCHeader(content: string, filePath: string): void {
     if (detectCppSyntax(content)) {
+      // Issue #211: C++ detected, set flag for .cpp output
+      this.cppDetected = true;
       // Use C++14 parser for headers with C++ syntax (typed enums, classes, etc.)
       this.parseCppHeader(content, filePath);
     } else {
@@ -580,7 +602,8 @@ class Pipeline {
    * Get output path for a file
    */
   private getOutputPath(file: IDiscoveredFile): string {
-    const ext = this.config.outputExtension;
+    // Issue #211: Derive extension from cppDetected flag
+    const ext = this.cppDetected ? ".cpp" : ".c";
     const outputName = basename(file.path).replace(/\.cnx$|\.cnext$/, ext);
 
     // Check if file is in any input directory (for preserving structure)

--- a/src/pipeline/types/IPipelineConfig.ts
+++ b/src/pipeline/types/IPipelineConfig.ts
@@ -23,8 +23,8 @@ interface IPipelineConfig {
   /** Whether to generate .h files for exported symbols (default: true) */
   generateHeaders?: boolean;
 
-  /** Output file extension (default: ".c") */
-  outputExtension?: ".c" | ".cpp";
+  /** Issue #211: Force C++ output (--cpp flag). Auto-detection may also enable this. */
+  cppRequired?: boolean;
 
   /** Parse only mode - no code generation */
   parseOnly?: boolean;

--- a/src/project/Project.ts
+++ b/src/project/Project.ts
@@ -49,7 +49,7 @@ class Project {
       defines: this.config.defines,
       preprocess: this.config.preprocess,
       generateHeaders: this.config.generateHeaders,
-      outputExtension: this.config.outputExtension,
+      cppRequired: this.config.cppRequired,
       noCache: this.config.noCache,
     });
   }

--- a/src/project/types/IProjectConfig.ts
+++ b/src/project/types/IProjectConfig.ts
@@ -29,8 +29,8 @@ interface IProjectConfig {
   /** Additional preprocessor defines */
   defines?: Record<string, string | boolean>;
 
-  /** Output file extension (default: ".c") */
-  outputExtension?: ".c" | ".cpp";
+  /** Issue #211: Force C++ output. Auto-detection may also enable this. */
+  cppRequired?: boolean;
 
   /** Issue #183: Disable symbol caching (default: false = cache enabled) */
   noCache?: boolean;

